cmd/soroban-rpc/internal/db/transaction.go

```diff
@@ -0,0 +1,336 @@
+// methods/get_transaction.go#NewGetTransactionHandler so that it can be used
+// directly against the RPC handler.
+//
+// Errors occur if there are issues with the DB connection or the XDR is
+// corrupted somehow. If the transaction is not found, io.EOF is returned.
+func (txn *transactionHandler) GetTransaction(ctx context.Context, hash xdr.Hash) (
+	Transaction, ledgerbucketwindow.LedgerRange, error,
+) {
+	start := time.Now()
+	tx := Transaction{}
+
+	ledgerRange, err := txn.GetLedgerRange(ctx)
+	if err != nil && err != ErrEmptyDB {
+		return tx, ledgerRange, err
+	}
+
+	lcm, ingestTx, err := txn.getTransactionByHash(ctx, hash)
+	if err != nil {
+		return tx, ledgerRange, err
+	}
+	tx, err = ParseTransaction(lcm, ingestTx)
+	if err != nil {
+		return tx, ledgerRange, err
+	}
+
+	txn.log.
+		WithField("txhash", hex.EncodeToString(hash[:])).
+		WithField("duration", time.Since(start)).
+		Debugf("Fetched and encoded transaction from ledger %d", lcm.LedgerSequence())
+
+	return tx, ledgerRange, nil
+}
+
+// getTransactionByHash actually performs the DB ops to cross-reference a
+// transaction hash with a particular set of ledger close meta and parses out
+// the relevant transaction efficiently by leveraging the `application_order` db
+// field.
+//
+// Note: Caller must do input sanitization on the hash.
+func (txn *transactionHandler) getTransactionByHash(ctx context.Context, hash xdr.Hash) (
+	xdr.LedgerCloseMeta, ingest.LedgerTransaction, error,
+) {
+	rowQ := sq.Select("t.application_order", "lcm.meta").
+		From(fmt.Sprintf("%s t", transactionTableName)).
+		Join(fmt.Sprintf("%s lcm ON (t.ledger_sequence = lcm.sequence)", ledgerCloseMetaTableName)).
+		Where(sq.Eq{"t.hash": []byte(hash[:])}).
+		Limit(1)
+
+	rows, err := txn.db.Query(ctx, rowQ)
+	if err != nil {
+		return xdr.LedgerCloseMeta{}, ingest.LedgerTransaction{},
+			errors.Wrapf(err, "failed row read for txhash %s", hash)
+	}
+	if !rows.Next() {
+		if ierr := rows.Err(); ierr != nil {
+			return xdr.LedgerCloseMeta{}, ingest.LedgerTransaction{},
+				errors.Wrapf(ierr, "failed to fetch txhash %s", hash)
+		}
+
+		// We use EOF to mark "this tx doesn't exist," since that isn't a real error.
+		return xdr.LedgerCloseMeta{}, ingest.LedgerTransaction{}, io.EOF
+	}
+
+	var (
+		txIndex int
+		lcm     xdr.LedgerCloseMeta
+	)
+	if err := rows.Scan(&txIndex, &lcm); err != nil {
+		return xdr.LedgerCloseMeta{}, ingest.LedgerTransaction{},
+			errors.Wrapf(err, "db read failed for txhash %s", hash)
+	}
+
+	reader, err := ingest.NewLedgerTransactionReaderFromLedgerCloseMeta(txn.passphrase, lcm)
+	reader.Seek(txIndex - 1)
+	if err != nil {
+		return lcm, ingest.LedgerTransaction{},
+			errors.Wrapf(err, "failed to index to tx %d in ledger %d (txhash=%s)",
+				txIndex, lcm.LedgerSequence(), hash)
+	}
+
+	ledgerTx, err := reader.Read()
+	return lcm, ledgerTx, err
+}
+
+func ParseTransaction(lcm xdr.LedgerCloseMeta, ingestTx ingest.LedgerTransaction) (Transaction, error) {
```

Comment:
is this function called outside the db package? if not, then I don't think we need to export it

Reply:
The intent was for @aditya1702 to use it in a later refactor of [this code diff](https://github.com/stellar/soroban-rpc/pull/136/files#diff-5276db8001730f32b267cab637b316958b69aef6f8184343338855fdc118b700R211), i.e. replacing `transactionsRPCHandler.getTransactionDetails(tx ingest.LedgerTransaction) ([]byte, []byte, []byte, error)` which is doing the same stuff.